pystac_client/client.py

```diff
@@ -504,3 +504,40 @@ def get_search_link(self) -> Optional[pystac.Link]:
             ),
             None,
         )
+
+    def _get_collections_href(self, id: Optional[str] = None) -> str:
+        self_href = self.get_self_href()
+        if self_href is None:
+            data_link = self.get_single_link("data")
+            if data_link is None:
+                raise ValueError(
+                    "cannot build a collections href without a self href or a data link"
+                )
+            else:
+                collections_href = data_link.href
```

Comment:
Is it probable that the data_link also ends with "/"?

Reply:
Yes, but that should be captured below: `elif collections_href.endswith("/"):`, right?

Reply:
If the id is None, it's returning the href with /

Reply:
Which is the right behavior. The client should use what the back-end reports. It may intentionally add a slash because it needs it. Forcing to remove it may result in issues with some back-ends.

Reply:
But then, does it not also affect how we construct the collection URLs L518 onwards? For example, you might want to have / behind `collections` the self_href ended with /.

Reply:
Yes, indeed. I wouldn't directly assume that a slash at the end of the self link would also require a slash for collections, but it should append one if the data link has a slash. Or better: Use the data link as default, but it seems @gadomski had reasons against that?

Reply:
Given the API specification itself is HAETEOS compliant, I would also recommend using the links (eg. data link) and craft URL as a fallback. But again [specification](https://github.com/radiantearth/stac-api-spec/tree/main/ogcapi-features) asks for URL without a slash. Do you think it shouldn't be rather enforced?

Reply:
I don't think I'd read the spec in a way that it doesn't allow slashes at the end...

Reply:
> Or better: Use the data link as default, but it seems @gadomski had reasons against that?

I don't have anything against it _per se_, but the current behavior of this library is to construct the collections URLs as strings from the specification (rather than look them up via the links), and so I kept that behavior as the default.

Note that the `/search` url is constructed from the link: https://github.com/stac-utils/pystac-client/blob/a5705e0f4bc1ff2cd80b8141310a303a77acbba4/pystac_client/client.py#L487-L506

IMO url construction in `Client` should be consistent, and probably configurable (see https://github.com/stac-utils/pystac-client/issues/320#issuecomment-1255014811). IMO that's a larger effort than this PR, but one worth tackling (I'll take a crack at it). This PR does not block or otherwise prevent us from later bringing things in line, so I'm going to leave it open.

Reply:
Indeed, the same applies for search, items, etc.